core/src/main/java/com/linecorp/armeria/common/HttpRequest.java

```diff
@@ -605,4 +606,66 @@ default HttpRequest mapError(Function<? super Throwable, ? extends Throwable> fu
         requireNonNull(function, "function");
         return of(headers(), HttpMessage.super.mapError(function));
     }
+
+    /**
+     * Applies the specified {@link Consumer} to the {@link HttpData}s
+     * emitted by this {@link HttpRequest}.
+     *
+     * <p>For example:<pre>{@code
+     * HttpRequest request = HttpRequest.of(RequestHeaders.of(HttpMethod.POST, "/items"),
+     *                                      HttpData.ofUtf8("data1,data2"));
+     * HttpRequest result = request.peekData(data -> {
+     *     assert data.toStringUtf8().equals("data1,data2");
+     * });
+     * }</pre>
+     */
+    default HttpRequest peekData(Consumer<? super HttpData> action) {
+        requireNonNull(action, "action");
+        final StreamMessage<HttpObject> stream = peek(action, HttpData.class);
+        return of(headers(), stream);
+    }
+
+    /**
+     * Applies the specified {@link Consumer} to the {@linkplain HttpHeaders trailers}
+     * emitted by this {@link HttpRequest}.
+     *
+     * <p>For example:<pre>{@code
+     * HttpRequest request = HttpRequest.of(RequestHeaders.of(HttpMethod.POST, "/items"),
+     *                                      HttpData.ofUtf8("..."));
+     * HttpRequest transformed = request
+     *     .mapTrailers(trailers -> {
+     *         return trailers.withMutations(builder -> builder.set("trailer1", "foo"));
+     *     })
+     *     .peekTrailers(trailers -> {
+     *         assert trailers.get("trailer1").equals("foo");
+     *     });
+     * }</pre>
+     */
+    default HttpRequest peekTrailers(Consumer<? super HttpHeaders> action) {
+        requireNonNull(action, "action");
+        final StreamMessage<HttpObject> stream = peek(obj -> {
+            if (!(obj instanceof ResponseHeaders)) {
```

Comment:
I guess we don't need this check. It always `HttpHeaders` that this request stream message has.